sdk/core/azure-core/src/io/body_stream.cpp

```diff
@@ -82,35 +83,120 @@ int64_t MemoryBodyStream::OnRead(Context const& context, uint8_t* buffer, int64_
   return copy_length;
 }
 
-#if defined(AZ_PLATFORM_POSIX)
+int64_t FileBodyStream::GetFileSize(FILE* file)
+{
+  // Get the current file position, to reset it back, after seeking to the end.
+  fpos_t currentPosition;
+  if (fgetpos(file, &currentPosition))
+  {
+    throw std::runtime_error("Failed to get the file object position.");
+  }
+
+  if (fseek(file, 0, SEEK_END))
+  {
+    throw std::runtime_error("Failed to seek to the end of the file.");
+  }
+
+  auto fileSize = ftell(file);
+  if (fileSize == -1)
+  {
+    throw std::runtime_error("Failed to get the size of the file.");
+  }
+
+  // Reset the file position back to what it was originally set to.
+  if (fsetpos(file, &currentPosition))
+  {
+    throw std::runtime_error("Failed to set the file object position.");
+  }
+
+  return fileSize;
+}
+
+FileBodyStream::FileBodyStream(FILE* file, int64_t offset, int64_t length)
+    : m_fileStream(file), m_baseOffset(offset), m_length(length), m_offset(0)
+{
+  if (file == NULL)
+  {
+    throw std::invalid_argument(
+        "The file object cannot be null and must have been successfully opened.");
+  }
+
+  if (offset < 0 || length < 0)
+  {
+    throw std::invalid_argument("The file offset and size must be a non-negative number.");
+  }
+
+  int64_t fileSize = GetFileSize(m_fileStream);
+  if (offset > fileSize || length > fileSize || offset + length > fileSize)
+  {
+    throw std::invalid_argument("The offset and length cannot be larger than the file size.");
+  }
+
+#if defined(AZ_PLATFORM_WINDOWS)
+  m_fileDescriptor = _fileno(m_fileStream);
+  m_filehandle = (HANDLE)_get_osfhandle(m_fileDescriptor);
+#elif defined(AZ_PLATFORM_POSIX)
+  m_fileDescriptor = fileno(m_fileStream);
+#endif
+}
+
+FileBodyStream::FileBodyStream(FILE* file, int64_t offset)
```

Comment:
I'm so confused as to why all the FILE* stuff is still here. 
I thought we agreed that we only support OS file constructs because we want BodyStream ot be FAST. We get this by calling ReadFileEx and pread; both of these functions REQUIRE a handle/fd and do NOT work with a FILE*.

Reply:
> I thought we agreed that we only support OS file constructs because we want BodyStream ot be FAST. We get this by calling ReadFileEx and pread; both of these functions REQUIRE a handle/fd and do NOT work with a FILE*.

The implementation of `OnRead` continues to use pread/ReadFileEx and we get the performance. I am just making handle/fd private impl rather than public API.